src/asmcnc/apps/upgrade_app/screen_upgrade.py

```diff
@@ -0,0 +1,268 @@
+        self.l = kwargs['localization']
+
+    def on_pre_enter(self):
+        # Reset app
+        self.update_strings()
+        self.hide_error_message()
+        self.m.write_dollar_51_setting(1)
+
+    def quit_to_lobby(self):
+        self.sm.current = 'lobby'
+        self.m.write_dollar_51_setting(0)
+
+    def get_correct_unlock_code(self, serial):
+        try:
+            return str(hex((serial + 42) * 10000))[2:]
+        except TypeError:
+            return 1
+
+    def code_entered(self):
+        self.hide_error_message()
+        self.show_verifying()
+        self.m.s.write_command('M3 S0')
+        Clock.schedule_once(self.get_restore_info, 0.1)
+
+    def get_restore_info(self, dt):
+        self.m.s.write_protocol(self.m.p.GetDigitalSpindleInfo(), "GET DIGITAL SPINDLE INFO")
+        self.check_info_count = 0
+        Clock.schedule_once(self.check_restore_info, 0.3)
+
+    def check_restore_info(self, dt):
+        self.check_info_count += 1
+        # Value of -999 represents disconnected spindle - if detected then stop waiting
+        if (self.m.s.digital_spindle_ld_qdA != -999 and self.m.s.spindle_serial_number not in [None, -999, 999]) or (self.check_info_count > 10):
+            self.read_restore_info()
+        else: # Keep trying for a few seconds
+            Clock.schedule_once(self.check_restore_info, 0.3)
+
+    def read_restore_info(self):
+        self.m.s.write_command('M5')
+        self.hide_verifying()
+        # Value of -999 for ld_qdA represents disconnected spindle
+        if self.m.s.digital_spindle_ld_qdA != -999 and self.m.s.spindle_serial_number not in [None, -999, 999]:
+            # Get info was successful, show serial and check code
+            self.spindle_label.text = (
+                self.l.get_str("Need support?") + " " + \
+                self.l.get_str("Quote Spindle motor number: NN").replace("NN", str(self.m.s.spindle_serial_number))
+            )
+            self.check_unlock_code()
+        else:
+            # Otherwise, spindle is probably disconnected
+            self.show_error_message(self.l.get_str("No SC2 Spindle motor detected.") + " " + self.l.get_str("Please check your connections."))
+            self.spindle_label.text = self.l.get_str("Need support?") + " " + self.l.get_str('Quote "No SC2"')
+
+    def check_unlock_code(self):
+        correct_unlock_code = self.get_correct_unlock_code(self.m.s.spindle_serial_number)
+        entered_unlock_code = self.upgrade_code_input.text.lower().replace('o', '0')
+
+        if correct_unlock_code == entered_unlock_code:
+            self.upgrade_and_proceed()
+        else:
+            self.show_error_message(self.l.get_str("Upgrade code incorrect, please check it and try again."))
+
+    def upgrade_and_proceed(self):
+        try:
+            self.m.enable_theateam()
+            self.sm.current = "upgrade_successful"
+        except:
+            warning_message = 'Problem creating SC2 compatability file!!'
```

Comment:
Not translated and reveals info about the unlock process onscreen, can you change to generic "Error!" and then print this error message so we can see it in logs if we need :)